### PR TITLE
Support toluene solvent

### LINKED ISSUE #222: add toluene as solvent
For SAMPL9, we want to evaluate solvation free energies in [toluene](https://pubchem.ncbi.nlm.nih.gov/compound/1140).

Toluene_300

- [x] add force field files AMBER GAFF
- [ ] add force field files CGENFF
- [x] add force field files OPLS-AA (mol2ff)
- [ ] integrate in the solvent selections 
- [ ] add basic tests

In general, follow what we did for cyclohexane.

### PR DESCRIPTION
**Pull request by VOD555**
Fix #222 

- [x] Add solvent model option in `forcefields.py`
- [x] Add classes for equilibrium and fep simulations to `equil.py`, `fep.py` and `run.py`
- [x] For AMBER and OPLS-AA
- [ ] For CAHRMM
- [ ] Support toluene in scripts
- [ ] Tests

### CODE REVIEW

File: mdpow/forcefields.py [hunk 1]
```diff
@@ -192,6 +194,8 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
         identifier="wetoctanol", itp="1octwet.itp", coordinates="1octwet_amber.gro"),
     'cyclohexane': GromacsSolventModel(
         identifier="cyclohexane", itp="1cyclo.itp", coordinates="1cyclo_amber.gro"),
+    'toluene': GromacsSolventModel(
+        identifier="toluene", itp="1tol.itp", coordinates="1tol_gaff.gro"),
```
Comment (iorga) [inline mdpow/forcefields.py:198]: I see that for the other solvents we used "amber" instead of "gaff", so the file should be renamed for consistency.

**Review by iorga — CHANGES_REQUESTED** [review]
The file `1tol_gaff.gro` should be renamed into `1tol_amber.gro`. I approve all other modifications.

File: mdpow/top/amber99sb.ff/1tol.itp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+     6   Ha     1   SOL   H11    6     0.130000      1.00800 ; qtot 0.077
+     7   Ca     1   SOL    C4    7    -0.127000     12.01000 ; qtot -0.050
+     8   Ha     1   SOL   H12    8     0.130000      1.00800 ; qtot 0.080
+     9   Ca     1   SOL    C5    9    -0.134999     12.01000 ; qtot -0.055
+    10   Ha     1   SOL   H13   10     0.130000      1.00800 ; qtot 0.075
+    11   Ca     1   SOL    C6   11    -0.127000     12.01000 ; qtot -0.052
+    12   Ha     1   SOL   H14   12     0.130000      1.00800 ; qtot 0.078
+    13   Ca     1   SOL    C7   13    -0.131000     12.01000 ; qtot -0.053
+    14   Ha     1   SOL   H15   14     0.130000      1.00800 ; qtot 0.077
+    15   Ca     1   SOL    C2   15    -0.077300     12.01000 ; qtot -0.000
```
Comment (orbeckst) [inline mdpow/top/amber99sb.ff/1tol.itp:28]: @iorga do you agree with introducing these atom types with these specific names? 

We need to add them to the standard files somewhere but do you like the names?

Comment (orbeckst) [reply mdpow/top/amber99sb.ff/1tol.itp:28]: See `mdpow_solvent.itp`, which keeps them contained.

File: mdpow/top/amber99sb.ff/mdpow_solvent.itp [hunk 3]
```diff
@@ -0,0 +1,7 @@
+[ atomtypes ]
+; name      at.num  mass     charge ptype  sigma      epsilon
+; atomtype for 1tol.itp
+ C3       0          0.0000  0.0000  A   3.39771e-01   4.51035e-01 ; 1.91  0.1078
+ Hc       0          0.0000  0.0000  A   2.60018e-01   8.70272e-02 ; 1.46  0.0208
+ Ca       0          0.0000  0.0000  A   3.31521e-01   4.13379e-01 ; 1.86  0.0988
+ Ha       0          0.0000  0.0000  A   2.62548e-01   6.73624e-02 ; 1.47  0.0161
```
Comment (orbeckst) [inline mdpow/top/amber99sb.ff/mdpow_solvent.itp:7]: possibly add newline?

File: mdpow/top/amber99sb.ff/ffnonbonded.itp [hunk 4]
```diff
@@ -72,4 +72,4 @@ OW_tip5p     8      16.00    0.0000  A   3.12000e-01  6.69440e-01
 MW           0       0.0000  0.0000  D   0.00000e+00  0.00000e+00
 ; Dummy masses for rigid CH3 and NH3 groups
 MCH3         0       0.0000  0.0000  A   0.00000e+00  0.00000e+00
-MNH3         0       0.0000  0.0000  A   0.00000e+00  0.00000e+00
+MNH3         0       0.0000  0.0000  A   0.00000e+00  0.00000e+00
```
Comment (orbeckst) [inline mdpow/top/amber99sb.ff/ffnonbonded.itp:75]: newline?

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looking good. Tests could be more extensive but as long as you added the same ones that we have for cyclohexane, it's ok.

Please add an entry to CHANGES.

File: mdpow/top/amber99sb.ff/1tol.itp (hunk 2 above)
Comment (iorga) [reply mdpow/top/amber99sb.ff/1tol.itp:28]: You are right @orbeckst , I think we should replace the atom types by the standard ones, it seems that we did that for cyclohexane and octanol. I'll try to look into it tomorrow.

Comment (orbeckst) [reply mdpow/top/amber99sb.ff/1tol.itp:28]: I think @VOD555 found that the atomtypes in your toluene do not exactly exist with the same LJ parameters in the standard FF — or was it only the charges that differed??? In any case, can we proceed and fix it later?

Comment (VOD555) [reply mdpow/top/amber99sb.ff/1tol.itp:28]: @iorga Is there a gromacs itp file that contains all atomtypes in GAFF (https://ambermd.org/antechamber/gaff.html Table I (a))?

Comment (orbeckst) [reply mdpow/top/amber99sb.ff/1tol.itp:28]: From what Shujie is showing, we should ideally include a "gaff.itp" file and the reference those standard GAFF atom types. 

If we don't have that, then the current solution with `mdpow_solvent.itp` seems reasonable. I'd just make sure that the atmo type names are consistent with GAFF.

**Comment by VOD555** [comment]
@orbeckst @iorga The pull request is really for review.

**Review by orbeckst — APPROVED** [review]
This looks ok to me but I would appreciate @iorga ’s opinion, too.

File: mdpow/run.py [hunk 5]
```diff
@@ -277,19 +278,21 @@ def fep_simulation(cfg, solvent, **kwargs):
...
-        'cyclohexane':equil.CyclohexaneSimulation
+        'cyclohexane': equil.CyclohexaneSimulation,
+        'toluene': equil.TolueneSimulation,
         }
     Simulations = {
         'water': fep.Ghyd,
         'octanol': fep.Goct,
         'wetoctanol': fep.Gwoct,
-        'cyclohexane':fep.Gcyclo
+        'cyclohexane':fep.Gcyclo,
```
Comment (iorga) [inline mdpow/run.py:288]: Add a space before fep.Gcyclo for consistency.

File: scripts/mdpow-ptw [hunk 6]
```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/env python
+"""%(prog)s [options] DIRECTORY [DIRECTORY ...]
+
+Run the free energy analysis for water and toluene in <DIRECTORY>/FEP
+and return the toluene-water partition coefficient log P_cw.
```
Comment (iorga) [inline scripts/mdpow-ptw:5]: P_cw should be P_tw

File: scripts/mdpow-ptw [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+the MD FEP simulations. It relies on the canonical naming scheme (basically:
+just use the defaults as in the tutorial).
+
+The dV/dlambda plots can be produced automatically (--plot=auto). If multiple
+DIRECTORY arguments are provided then one has to choose the auto option (or
+None).
+
+Results are *appended* to a data file with **Output file format**::
+
+ molecule  DeltaG0(kJ/mol) ErrorDeltaG0  "logPtw"  log P_TW  Error logPcw   directory
```
Comment (iorga) [inline scripts/mdpow-ptw:18]: logPcw should be logPtw

File: scripts/mdpow-ptw [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+molecule
+    molecule name as used in the itp
+DeltaG0, ErrorDeltaG0
+    transfer free energy water --> toluene, in kJ/mol; error from correlation time
+    and error propagation through Simpson's rule (see :meth:`mdpow.fep.Gsolv`)
+log P_CW
```
Comment (iorga) [inline scripts/mdpow-ptw:25]: P_CW should be P_TW

File: scripts/mdpow-ptw [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+             append individual energy terms to this file
+        *force*
+             force rereading data files [False]
+
+    **Output file format (energies and log P_tw with errors) :**
+
+        molecule   DeltaG0(kJ/mol)    log P_TW    directory
+
+    Removed from output (>0.2.1):
+     * wat_ok and cyc_ok are "OK" if no lines were skipped in the input data files due to
```
Comment (iorga) [inline scripts/mdpow-ptw:84]: I think that cyc_ok should be tol_ok.

File: scripts/mdpow-ptw [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+       file corruption, and "BAD" otherwise.
+
+    """
+    import gc
+
+    fmt_result = "%-8s %+8.2f %8.2f  logPcw %+8.2f %8.2f   %s\n"
```
Comment (iorga) [inline scripts/mdpow-ptw:90]: logPcw should be logPtw

File: scripts/mdpow-ptw [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    fmt_energy = "%s  %+8.2f %8.2f   logPcw %+8.2f %8.2f  %s\n"
```
Comment (iorga) [inline scripts/mdpow-ptw:91]: logPcw should be logPtw

File: scripts/mdpow-ptw [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+        import matplotlib
+        from matplotlib.pyplot import clf, title, savefig
+        clf()
+        gwat.plot(color='k', ls='--')
+        gtol.plot(color='r', ls='-')
+        activate_subplot(1)
+        title(r"[%s] $\Delta A^0 = %.2f\pm%.2f$ kJ/mol" % \
+                  (gtol.molecule, transferFE.value, transferFE.error))
+        activate_subplot(2)
+        title(r"$\log P_{CW} = %.2f\pm%.2f$" % logPcw.astuple())
```
Comment (iorga) [inline scripts/mdpow-ptw:144]: P_{CW} should be P_{TW} and logPcw should be logPtw

File: scripts/mdpow-ptw [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+    parser = argparse.ArgumentParser(usage=__doc__, prog='mdpow-ptw',
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument("directory", nargs='+',
+                        help="directory or directories which contain all the files "
+                        "resulting from running mdpow.fep.Ghyd.setup() and mdpow.fep.Gtol.setup()"
+                        "and the results of the MD FEP simulations.",)
+    parser.add_argument('--plotfile', dest="plotfile", default='none',
+                        help="plot dV/dlambda to FILE; use png or pdf suffix to "
+                        "determine the file type. 'auto' generates a pdf file "
+                        "DIRECTORY/dVdl_<molname>_pcw.pdf and 'none' disables it "
```
Comment (iorga) [inline scripts/mdpow-ptw:196]: _pcw should be _ptw

File: scripts/mdpow-ptw [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+    for directory in opts.directory:
+        if not os.path.exists(directory):
+            logger.warning("Directory %r not found, skipping...", directory)
+            continue
+        logger.info("Analyzing directory %r... (can take a while)", directory)
+
+        opts.plotfile = parse_filename_option(opts.plotfile)
+
+        try:
+            run_pcw(directory, plotfile=opts.plotfile,
```
Comment (iorga) [inline scripts/mdpow-ptw:258]: run_pcw should be run_ptw

**Review by iorga — COMMENTED** [review]
I'm OK with these changes, but there are some minor fixes (generally forgotten changes from cyclohexane to toluene) to be done before the final merge.

The implementation of toluene solvent in GAFF looks functional, at least for the initial steps, as I could already set up the equilibrium simulations witout errors.

**Review by iorga — APPROVED** [review]

**Comment by orbeckst** [comment]
Please squash-merge yourself whe you’re ready. Am 1/10/23 um 13:18 schrieb Shujie Fan ***@***.***>:﻿
It should be ready to be merged.
I'll make a separate pull-request after we test the cgenff parameters.

**Comment by VOD555** [comment]
@orbeckst I'm not able to merge, as the 2 coverage checks don't pass.

**Comment by orbeckst** [comment]
Ok, I used my admin powers.

**Pull request merged**